web/app.js

```diff
@@ -1133,12 +1133,18 @@ const PDFViewerApplication = {
     }
   },
 
-  downloadOrSave(options = {}) {
-    if (this.pdfDocument?.annotationStorage.size > 0) {
-      this.save(options);
-    } else {
-      this.download(options);
-    }
+  async downloadOrSave(options = {}) {
+    // In the Firefox case, this method MUST always trigger a download.
+    // When the user is closing a modified and unsaved document, we display a
+    // prompt asking for saving or not. In case they save, we must wait for
+    // saving to complete before closing the tab.
+    // So in case this function does not trigger a download, we must trigger a
+    // a message and change PdfjsChild.sys.mjs to take it into account.
+    document.body.classList.add("wait");
```

Comment:
We already have a reference to this DOM element, from [here](https://github.com/mozilla/pdf.js/blob/47791a4c8088582d3f3374beab7e2af4d35153fd/web/viewer.js#L39), so we should use that in this method as well for consistency.
```suggestion
    const { classList } = this.appConfig.appContainer;
    classList.add("wait");
```